### PR TITLE
Extend Ivy Functional API  huber_loss

### LINKED ISSUE #21069: huber_loss
- [huber_loss] #22375

### PR DESCRIPTION
**Pull request by sushmanthreddy**
close #21069

### CODE REVIEW

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @sushmanthreddy ! Thanks for opening this PR! Would you please add a container and array method  in `ivy/container/experimental/losses.py` and the array instance method in `ivy/array/experimental/losses.py`. Feel free to request another review when you are done.

**Comment by sushmanthreddy** [comment]
> Hey @sushmanthreddy ! Thanks for opening this PR! Would you please add a container and array method in `ivy/data-classes/container/experimental/losses.py` and `ivy/data-classes/array/experimental/losses.py` respectively. Feel free to request another review when you are done.

i have added the container and array static methods..
pls once review it and merge it

File: ivy/functional/ivy/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/losses.py:434]: There's no function `assert_float_param` in `assertions`. I am not sure what exactly you are trying to do here.  Could you please explain this?

**Comment by hello-fri-end** [comment]
ivy-gardener

**Comment by sushmanthreddy** [comment]
@hello-fri-end can u pls review the pr ?i have made necessary changes

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:109]:
```suggestion
    delta: Optional[float] = 1.0,
    reduction: Optional[str] = "mean",
```

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:150]: Tbh, I don't think there's any need to add this since we have already mentioned in the type-hints that `delta` should be a float.

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:12]:
``` suggestion
       pred: Union[ivy.Array, ivy.NativeArray],
```

Just to be consistent.

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:106]:
```suggestion
    true: Union[ivy.Array, ivy.NativeArray],
    pred: Union[ivy.Array, ivy.NativeArray],
```
To be consistent with other loss functions.

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:149]:
```suggestion
        try:
        assert true.shape == pred.shape
    except ValueError:
        raise ValueError(
            "`pred` and `true` must have the same shape, received "
            f"({pred.shape} vs {true.shape})."
        )
```

We should be asserting that the shapes are equal.

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:119]: We should changes these to `true` and `pred` as well.

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:157]:
```suggestion
        ret = ivy.mean(loss, out=out)
```

So that the `out` parameter is also updated.

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:159]:
```suggestion
        ret = ivy.sum(loss, out=out)
```

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @sushmanthreddy ! Left a few more comments :)

Also, since pytorch has a functional implementation for `huber_loss` (https://pytorch.org/docs/stable/generated/torch.nn.functional.huber_loss.html), this should be implemented as a [mixed function](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions) and you should a add primary implementation for torch backend in `ivy/functional/backends/torch/experimental/losses.py`. Let me know if I should clarify anything further. Feel free to request another review when you are done with the changes, Thanks and happy coding :)

**Comment by sushmanthreddy** [comment]
@hello-fri-end sorry for so many commits , i have changed it on locally,


are there any required changes..to be made?

File: ivy/functional/ivy/experimental/losses.py
Comment (sushmanthreddy) [reply ivy/functional/ivy/experimental/losses.py]: yeah i thought of mentioning that it should be float only, anyways removed it..

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:14]:
```suggestion
        delta: Optional[float] = 1.0,
```

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:26]:
```suggestion
        pred
```

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:44]: let's change these to `true` and `pred`

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:48]:
```suggestion
        return ivy.huber_loss(self._data, pred, delta=delta, out=out)
```

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:16]:
```suggestion
        delta: Union[float] = 1.0,
```

Only arrays arguments can be containers.

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:100]:
```suggestion
        delta: Union[float] = 1.0,
```

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:158]:
```suggestion
    abs_diff = ivy.abs(true - pred)
```

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Have left a few more comments for you to resolve. Also, can you add a backend implementation for torch as I mentioned in the last review. Feel free to request another review when you are done. Thanks :)

**Comment by sushmanthreddy** [comment]
@hello-fri-end where , should i add in torch module there is no file called losses in torch..

I am new to ivy , I really don't knew about torch backend implementation, if you can show me a little example ,I can replicate that for huberloss

**Comment by hello-fri-end** [comment]
> @hello-fri-end where , should i add in torch module there is no file called losses in torch..
The function needs to be added here :`ivy/functional/backends/torch/experimental/losses.py`. You can create the  `losses.py` file if it does not already exist.

> I am new to ivy , I really don't knew about torch backend implementation, if you can show me a little example ,I can replicate that for huberloss
You can take a look at https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions for understanding mixed functions. To summarise though, the function signature in the torch backend implementation should be the same as in the functional implementation and you should use the `torch.nn.functional.huber_loss` function in the implementation. Let me know if you need me to clarify anything further. Thanks :)

**Comment by sushmanthreddy** [comment]
@hello-fri-end loss functions are composite functions . so it doesn't need backend implementation, I think we can implement in stateful function.
Screenshot 2023-08-03 at 10 55 51 AM

**Comment by hello-fri-end** [comment]
> @hello-fri-end loss functions are composite functions . so it doesn't need backend implementation, I think we can implement in stateful function. Screenshot 2023-08-03 at 10 55 51 AM

If there exists a native implementation for a function within any framework, we always add it to our backends. This is because the native implementations are mostly written in C++ which makes them more efficient than having a compositional implementation in Ivy. In this case, since torch has a `huber_loss` function, we should add it in our torch backend implementations. Hope that makes sense. Also, please go through the deep-dive I shared above for more details.

**Comment by sushmanthreddy** [comment]
@hello-fri-end backend in torch is implemented 
i have used l1_loss

**Comment by hello-fri-end** [comment]
> @hello-fri-end backend in torch is implemented i have used l1_loss
> 
> @hello-fri-end pls give a look on this reply once [link](https://github.com/unifyai/ivy/pull/21070#discussion_r1282567900)

You should implement huber_loss in the backend and not l1_loss.

**Comment by sushmanthreddy** [comment]
@hello-fri-end sorry i was working on l1_loss also..got confused without seeing

**Comment by sushmanthreddy** [comment]
@hello-fri-end sorry for the mistake i have use torch.nn.HuberLoss function and updated the backend implementation

**Comment by sushmanthreddy** [comment]
@hello-fri-end can u pls review again?I have added the huber loss backen implementation in torch

File: ivy/functional/backends/torch/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/backends/torch/experimental/losses.py:4]: Let's remove this since it's not used.
```suggestion
```

File: ivy/functional/backends/torch/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/backends/torch/experimental/losses.py:21]: You should be using `torch.nn.functional.huber_loss` here.Look at the documentation here https://pytorch.org/docs/stable/generated/torch.nn.functional.huber_loss.html and update the function call accordingly.

File: ivy/functional/backends/torch/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/backends/torch/experimental/losses.py:12]: You should run this script to check data-types supported by this function:
```python
import torch

def get_supported_dtypes():
    supported_dtypes = []
    unsupported_dtypes = []
    
    # Define all supported data types
    data_types = [torch.float16, torch.float32, torch.float64,
                  torch.bfloat16, torch.uint8, torch.int8,
                  torch.int16, torch.int32, torch.int64, torch.bool]
    
    for dtype in data_types:
        try:
            # Sample input tensors
            y_true = torch.tensor([1.0, 2.0, 3.0], dtype=dtype)
            y_pred = torch.tensor([1.1, 1.8, 2.9], dtype=dtype)
            
            # Check if the function raises an exception
            torch.nn.functional.huber_loss(y_true, y_pred)
            
            supported_dtypes.append(dtype)
        except :
            unsupported_dtypes.append(dtype)
    
    return supported_dtypes, unsupported_dtypes

if __name__ == "__main__":
    supported_dtypes, unsupported_dtypes = get_supported_dtypes()
    
    print("Supported dtypes:", supported_dtypes)
    print("Unsupported dtypes:", unsupported_dtypes)
   ```

It will return 

```python
Supported dtypes: [torch.float16, torch.float32, torch.float64, torch.bfloat16]
Unsupported dtypes: [torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64, torch.bool]
```

Please update the decorator accordingly.

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Please resolve these comments. Make sure that the tests are passing locally first. Feel free to request another review when you are done. Happy coding :)

File: ivy/functional/ivy/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/losses.py:386]:
```suggestion
huber_loss.mixed_backend_wrappers = {
    "to_add": (
        "handle_out_argument",
        "inputs_to_native_arrays",
        "outputs_to_ivy_arrays",
        "handle_device_shifting",
    ),
    "to_skip": ("inputs_to_ivy_arrays"),
}
```
Since it's a mixed function you need to add this.

**Comment by sushmanthreddy** [comment]
@hello-fri-end I am still confused ,why u mentioned me to write huberloss as mixed function ,because it is only supported by torch and tensor flow ...same like binary cross entropy and backend is not implemented for binary cross entrpy loss..

so , by taking example have implemented huber loss composite function.

Now test cases will pass ,only for paddle they will not  pass because of bfloat bug.
And u can see the bfloat bug here in this discussion. [link](https://discord.com/channels/799879767196958751/1139168214723592264)


And stateful API also has bug due to that we get an error `modulenotfound`.
And tests will not pass not only for huber loss which is implemented above but also for sparse_cross_entropy .

**Comment by sushmanthreddy** [comment]
@hello-fri-end I have implemented the huber_loss for above function and now tests are passing ,i have added the backend support also,can u pls review the pull request?

The paddle test is failing because of an error with the bfloat16 datatype which again is a known issue that Ivy people are working on.

so pls review it once again.

i hope will get those magical words "LGTM"..lol

**Comment by sushmanthreddy** [comment]
Screenshot from 2023-08-18 02-25-37
tests passing sc

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:15]:
```suggestion
        reduction: Optional[str] = "mean",
```

I believe this parameter should also be optional.

File: ivy/data_classes/array/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/losses.py:32]: Can you add `reduction` parameter here also?

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:16]:
```suggestion
        delta: Optional[Union[float, ivy.Container]] = 1.0,
```

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:17]:
```suggestion
        reduction: Optional[Union[str, ivy.Container]] = "mean",
```

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:37]: `reduction` here aswell./

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:103]:
```suggestion
        reduction: Optional[Union[str, ivy.Container]] = "mean",
```

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:102]:
```suggestion
        delta: Optional[Union[float, ivy.Container]] = 1.0,
```

File: ivy/data_classes/container/experimental/losses.py
Comment (hello-fri-end) [inline ivy/data_classes/container/experimental/losses.py:124]: `reduction` here.

File: ivy/functional/backends/paddle/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/backends/paddle/experimental/losses.py:35]: The function signature should match in all the implementations.

```python
def huber_loss(
    true: Union[paddle.tensor],
    pred: Union[paddle.tensor],
    /,
    *,
    delta: Optional[float] = 1.0,
    reduction: Optional[str] = "mean",
    out: Optionalpaddle.tensor] = None,
```

File: ivy/functional/backends/torch/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/backends/torch/experimental/losses.py:12]: Again, the function signature should match. These should be renamed to `true` and `pred`

File: ivy/functional/ivy/experimental/losses.py
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/losses.py:161]: I'm not sure if I follow this completely. `reduction` can only have two values `sum` or `mean`, so, why is there an else case here?

**Review by hello-fri-end — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/paddle/experimental/losses.py
Comment (sushmanthreddy) [reply ivy/functional/backends/paddle/experimental/losses.py:35]: @hello-fri-end can u plss see function implementation here , in huber loss and it doesn't have the reduction parameter in its backend ,that is the reason ,I have left out reduction over here..

Screenshot 2023-08-19 at 11 27 45 AM

File: ivy/functional/ivy/experimental/losses.py
Comment (sushmanthreddy) [reply ivy/functional/ivy/experimental/losses.py:161]: @hello-fri-end here the case where it is "none " is there actually none is also there for them implementation i that case else will work

**Comment by sushmanthreddy** [comment]
@hello-fri-end I have made necessary changes ,can u pls review it once again?if no changes are needed pls merge it..

**Pull request closed without merging**
